### PR TITLE
Implement Marking Lost Dog as Found (Backend)

### PR DESCRIPTION
**Pull request by nighthawkbravo**
- [x] Implement LostDogDataRepository (Estimate: 0.5h)
- [x] Implement LostDogService (Estimate: 1h)
- [x] Implement LostDogsController (Estimate: 1h)

### CODE REVIEW

File: Backend/src/main/java/se/backend/controller/internal/DogsController.java [hunk 1]
```diff
@@ -76,7 +76,7 @@ public DogsController(LostDogService lostDogService, LoginService loginService)
                                                         @RequestPart("dog") LostDogWithBehaviors newDog,
                                                         @RequestPart("picture") MultipartFile picture) {
         logHeaders(headers);
-        if(!loginService.IsAuthorized(headers, List.of(UserType.Regular))) {
+        if(!loginService.IsAuthorized(headers, List.of(UserType.Regular,))) {
```
Comment (Lank891) [inline Backend/src/main/java/se/backend/controller/internal/DogsController.java:79]: Does it compile? Tests are not passed and the error is that something in this file cannot be compiled

**Comment by nighthawkbravo** [comment]
Yes when I did a merge, my test for updating a dog went haywire. I am still debugging it.

File: Backend/src/main/java/se/backend/controller/internal/DogsController.java (hunk 1 above)
Comment (nighthawkbravo) [reply Backend/src/main/java/se/backend/controller/internal/DogsController.java:79]: Yes this error is fixed. I guess I didn't push it. But I have a different issue with the test in the dog controller that tests the updateDog function. I am still debugging it.

File: Backend/src/test/java/controller/internal/DogsControllerTest.java [hunk 2]
```diff
@@ -59,4 +60,74 @@ public void AddTest() throws Exception {
...
+                request.setMethod("PUT");
+                return request;
+            }
+        });
+        mockMvc.perform(builder
+                .file(dog)
+                .file(picture)
+                .header("token", "regularUserTestToken")
+        ).andExpect(status().isOk())
+                .andExpect(jsonPath("successful", is(true)));
```
Comment (Lank891) [inline Backend/src/test/java/controller/internal/DogsControllerTest.java:115]: Same applies here, check more stuff than only if code is 200, check if this is actually the dog you wanted to update

**Review by Lank891 — CHANGES_REQUESTED** [review]
Apart from the comments I think it's okay

File: Backend/src/main/java/se/backend/service/lostdogs/LostDogMainService.java [hunk 3]
```diff
@@ -74,4 +77,107 @@ public LostDogWithBehaviorsAndWithPicture AddLostDog(LostDogWithBehaviors newDog
...
+        pictureRepository.deleteById(dog.getPictureId());
+        lostDogRepository.delete(dog);
+        return true;
+    }
+
+    // method to check if the id exists in the database and if its not less then 0.
+    private boolean IsValidDogId(long dogId)
+    {
+        if(dogId < 0) return false;
+        var dogs = lostDogRepository.findAll(Specification.where(null), PageRequest.of(0, 15)).getContent();
```
[the remark this answers is not in the document]
Comment (nighthawkbravo) [reply Backend/src/main/java/se/backend/service/lostdogs/LostDogMainService.java]: I fixed the validator function to use "existById".

File: Backend/src/test/java/controller/internal/DogsControllerTest.java [hunk 4]
```diff
@@ -59,4 +60,74 @@ public void AddTest() throws Exception {
...
+                MockMvcRequestBuilders.get("/lostdogs/10001")
+                        .header("token", "regularUserTestToken")
+        ).andExpect(status().isOk())
+                .andExpect(jsonPath("successful", is(true)));
+
+        mockMvc.perform(
+                MockMvcRequestBuilders.get("/lostdogs/-1")
+                        .header("token", "regularUserTestToken")
+        ).andExpect(status().isBadRequest())
+                .andExpect(jsonPath("successful", is(false)));
```
[the remark this answers is not in the document]
Comment (nighthawkbravo) [reply Backend/src/test/java/controller/internal/DogsControllerTest.java:95]: I added more checking to the test.

File: Backend/src/test/java/controller/internal/DogsControllerTest.java (hunk 2 above)
Comment (nighthawkbravo) [reply Backend/src/test/java/controller/internal/DogsControllerTest.java]: Here as well.

File: Backend/src/main/java/se/backend/controller/internal/DogsController.java [hunk 5]
```diff
@@ -95,4 +96,69 @@ public DogsController(LostDogService lostDogService, LoginService loginService)
...
+
+    @DeleteMapping(path = "/{dogId}")
+    public ResponseEntity<Response<Boolean>> DeleteLostDog(@RequestHeader HttpHeaders headers, @PathVariable("dogId") long dogId) {
+        logHeaders(headers);
+        if(!loginService.IsAuthorized(headers, List.of(UserType.Admin, UserType.Regular))) {
+            throw new UnauthorizedException();
+        }
+
+        // If we successfully delete
+        if(lostDogService.DeleteDog(dogId)) return ResponseEntity.ok(new Response<>(String.format("Deleted dog with id: %d", dogId), true, true));
```
[the remark this answers is not in the document]
Comment (nighthawkbravo) [reply Backend/src/main/java/se/backend/controller/internal/DogsController.java:106]: I removed comments.

**Review by Lank891 — APPROVED** [review]

**Pull request merged**